catalog/dags/data_refresh/data_refresh_task_factory.py

```diff
@@ -152,25 +152,49 @@ def create_data_refresh_task_group(
         )
         tasks.append(generate_index_suffix)
 
-        action_data_map: dict[str, dict] = {
-            "ingest_upstream": {},
-            "promote": {"alias": target_alias},
-        }
-        for action, action_post_data in action_data_map.items():
-            with TaskGroup(group_id=action) as task_group:
-                ingestion_server.trigger_and_wait_for_task(
-                    action=action,
-                    model=data_refresh.media_type,
-                    data={
-                        "index_suffix": XCOM_PULL_TEMPLATE.format(
-                            generate_index_suffix.task_id, "return_value"
-                        ),
-                    }
-                    | action_post_data,
-                    timeout=data_refresh.data_refresh_timeout,
-                )
-
-            tasks.append(task_group)
+        # Trigger the 'ingest_upstream' task on the ingestion server and await its
+        # completion. This task copies the media table for the given model from the
+        # Catalog into the API DB and builds the elasticsearch index. The new table
+        # and index are not promoted until a later step.
+        with TaskGroup(group_id="ingest_upstream") as ingest_upstream_tasks:
+            ingestion_server.trigger_and_wait_for_task(
+                action="ingest_upstream",
+                model=data_refresh.media_type,
+                data={
+                    "index_suffix": XCOM_PULL_TEMPLATE.format(
+                        generate_index_suffix.task_id, "return_value"
+                    ),
+                },
+                timeout=data_refresh.data_refresh_timeout,
+            )
+            tasks.append(ingest_upstream_tasks)
+
+        # Poll the API, targeting the newly created index rather than the live index,
+        # and wait until healthy results are returned.
+        api_health_check = ingestion_server.api_health_check(
+            media_type=data_refresh.media_type,
+            index_suffix=XCOM_PULL_TEMPLATE.format(
+                generate_index_suffix.task_id, "return_value"
+            ),
+            timeout=data_refresh.api_healthcheck_timeout,
+        )
+        tasks.append(api_health_check)
+
+        # Trigger the `promote` task on the ingestion server and await its completion.
+        # This task promotes the newly created API DB table and elasticsearch index.
+        with TaskGroup(group_id="promote") as promote_tasks:
```

Comment:
Rather than duplicating this and splitting it out, is it maybe better to move it into a generation function that takes an action and returns the task group?

Reply:
Yeah, I went back and forth on this during implementation as well and I'm not sure! We could add another `trigger_and_wait_for_task_group` function that just wraps `trigger_and_wait_for_task` in a TaskGroup, but there's almost no code duplication saved because we still need to pass through all the same params. We can create the TaskGroups in a loop as was done originally, but the api health check needs to be between them in the task control flow. We _can_ set up the dependencies afterward but I thought doing it this way much improved readability, especially because it allowed me to more easily document the individual task groups in comments. I would definitely be more inclined to use the action map if we end up adding more than two trigger-and-wait taskgroups.

Reply:
Totally fair! We can keep it as is 😄